### PR TITLE
#4273 add setcookie $options for PHP 7.3 compatibility

### PR DESCRIPTION
**Pull request by thekonz**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #4273

### CODE REVIEW

**Comment by thekonz** [comment]
@samdark Any idea why this is failing? My test works on its own, but travis just cannot connect to the mysql server.

**Comment by My6UoT9** [comment]
According to https://github.com/travis-ci/travis-ci/issues/6842#issuecomment-502259411
> https://travis-ci.community/t/prs-started-failing-on-6-12/3707/2
>
>We are moving to Xenial as the default, and we do not start DB by default.
>
>https://blog.travis-ci.com/2019-04-15-xenial-default-build-environment#1-services-support
>
>You need to explicitly instruct that it is needed.
>
>services:
>  - mysql

https://github.com/yiisoft/yii/blob/master/.travis.yml
So maybe just add mysql to the services.
 This is strange since, the tests should be failing since a while - if this the reason?

**Comment by My6UoT9** [comment]
Seems mysql works now, but it wants psql now..
https://docs.travis-ci.com/user/database-setup/#postgresql
>services:
>  - postgresql

**Comment by My6UoT9** [comment]
Travis using xenial (`Build dist: xenial`)  instead of trusty.
Xenial seems to not support older php versions, trusty should.

I would try to revert the changes to services, and set dist to trusty instead of xenial (which is default).

https://docs.travis-ci.com/user/reference/trusty/
Would need this key in main (as seen in this example: https://docs.travis-ci.com/user/reference/trusty/#hhvm).
`dist: trusty`

**Comment by thekonz** [comment]
@My6UoT9 thank you :D

File: framework/web/CHttpCookie.php [hunk 1]
```diff
@@ -48,6 +48,11 @@ class CHttpCookie extends CComponent
 	 * Note, this property is only effective for PHP 5.2.0 or above.
 	 */
 	public $httpOnly=false;
+    /**
+     * @var array Additional setcookie options like "SameSite".
+     * This property only works for PHP 7.3.0 or above.
+     */
+	public $options=array();
```
Comment (samdark) [inline framework/web/CHttpCookie.php]: It seems there are no additional options other than SameSite. How about handing it the same way it's done in Yii 2: https://github.com/yiisoft/yii2/blob/master/framework/web/Cookie.php#L80 and Yii 3: https://github.com/yiisoft/yii-web/blob/master/src/Cookie.php#L79

File: framework/web/CHttpCookie.php [hunk 2]
```diff
@@ -48,6 +48,11 @@ class CHttpCookie extends CComponent
 	 * Note, this property is only effective for PHP 5.2.0 or above.
 	 */
 	public $httpOnly=false;
+    /**
+     * @var array Additional setcookie options like "SameSite".
```
Comment (samdark) [inline framework/web/CHttpCookie.php]: Formatting seems to be off. Not a huge issue though. Can fix it while merging.

File: framework/web/CHttpCookie.php (hunk 1 above)
Comment (thekonz) [reply framework/web/CHttpCookie.php]: > there are no additional options other than SameSite

... yet

I think that we want the code to be flexible for the future and prevent more pull requests for every other flag that browser vendors want to add.

File: framework/web/CHttpCookie.php (hunk 2 above)
Comment (thekonz) [reply framework/web/CHttpCookie.php]: Fixed it.

File: framework/web/CHttpCookie.php (hunk 1 above)
Comment (samdark) [reply framework/web/CHttpCookie.php]: That complicates the syntax. Additional properties could be added as needed.

Comment (thekonz) [reply framework/web/CHttpCookie.php]: > That complicates the syntax

That is a bold claim, but hey... have it your way.
I changed it.

**Comment by samdark** [comment]
Merged. Thank you!

**Comment by tomfotherby** [comment]
I propose to also update `setCookieParams`. I raised a PR: https://github.com/yiisoft/yii/pull/4304

**Pull request merged**